jax/interpreters/pxla.py

```diff
@@ -665,7 +742,21 @@ def _pval_to_result_handler(axis_size, nrep, pval, devices, backend):
                    else replicate(const, axis_size, nrep, devices, backend))
     return lambda _: bcast_const
   else:
-    return aval_to_result_handler(axis_size, nrep, pv)
+    if pv is not core.abstract_unit:
+      sharding_spec = _pmap_sharding_spec(nrep, axis_size, pv.shape)
```

Comment:
As long as the sharding spec is the same for all non-const outputs, we should probably use the copy constructed in parallel_callable? (While my changes will make the sharding spec no longer the same for all outputs, there will still be only two possibilities.)

Reply:
See other comment. I also suspect it's cheap constructing these compared to compilation cost (haven't measured though!).

Reply:
Yeah, also as you point out even if the sharding specs are the same the indices aren't because the shapes are different. The current implementation LGTM.

Reply:
Note sharding specs are specific to the array rank, so even the sharding specs can be different (len(shards_per_axis) == len(is_axis_materialized) == rank). I thought about using prefixes so different shapes/ranks could share sharding specs, but I'm not sure it'd actually buy you much due to these other concerns and would add some complexity to the implementation.